### DISCUSSION TITLE
Migrate Test Method With Generics

### DISCUSSION THREAD

**Discussion by viceroypenguin** [thomhurst/TUnit#945] [Q&A] [answered]
Current XUnit Code:
```cs
[Theory]
[MemberData(nameof(AggregateBy_TestData))]
public static void AggregateBy_HasExpectedOutput<TSource, TKey, TAccumulate>(
	IEnumerable<TSource> source,
	Func<TSource, TKey> keySelector,
	Func<TKey, TAccumulate> seedSelector,
	Func<TAccumulate, TSource, TAccumulate> func,
	IEqualityComparer<TKey>? comparer,
	IEnumerable<KeyValuePair<TKey, TAccumulate>> expected) where TKey : notnull
{
	using var ts = source.AsTestingSequence();
	SuperEnumerable
		.AggregateBy(ts, keySelector, seedSelector, func, comparer)
		.AssertCollectionEqual(expected);
}

public static IEnumerable<object?[]> AggregateBy_TestData()
{
	yield return WrapArgs(
		source: Enumerable.Range(0, 10),
		keySelector: x => x,
		seedSelector: x => 0,
		func: (x, y) => x + y,
		comparer: null,
		expected: Enumerable.Range(0, 10).ToDictionary(x => x, x => x));

	yield return WrapArgs(
		source: ["Bob", "bob", "tim", "Bob", "Tim"],
		keySelector: x => x,
		seedSelector: x => "",
		func: (x, y) => x + y,
		comparer: null,
		expected: new Dictionary<string, string>(StringComparer.Ordinal)
		{
			{ "Bob", "BobBob" },
			{ "bob", "bob" },
			{ "tim", "tim" },
			{ "Tim", "Tim" },
		});

	// .. other test cases

	static object?[] WrapArgs<TSource, TKey, TAccumulate>(IEnumerable<TSource> source, Func<TSource, TKey> keySelector, Func<TKey, TAccumulate> seedSelector, Func<TAccumulate, TSource, TAccumulate> func, IEqualityComparer<TKey>? comparer, IEnumerable<KeyValuePair<TKey, TAccumulate>> expected)
		=> [source, keySelector, seedSelector, func, comparer, expected];
}
```

What would be the best way to migrate this to TUnit `[MethodDataSource]`?

**Comment by thomhurst** [comment]
That is some complicated test haha, just trying to wrap my head around what it's doing!

I see you've raised an issue, is that to sort this?

I think extending method data source to allow inputs seems like a good idea so I'm happy to try and incorporate that

**Reply by viceroypenguin** [reply]
The issue is unrelated, and I found a better solution for my exact situation for that; just pointing it out for others who may have scenarios that would benefit from that improvement.

**Reply by viceroypenguin** [reply]
> That is some complicated test haha, just trying to wrap my head around what it's doing!

The goal is to simplify testing the `AggregateBy` method in order to not have to write the same code for different types, when it would look exactly the same. So the test method is generic, and the method returns values which are then used to specialize the test method, and then the values passed in for the test.

Having written this out, perhaps it's too much magic, and I'd be better off de-genericizing the method, actually. Unless you come up with a tricky idea, that is. Haha. :)

**Reply by thomhurst** [reply]
Did you try this in TUnit with MethodDataSource? If it doesn't work for whatever reason let me know what issue you're getting and let's see if we can solve it

**Reply by viceroypenguin** [reply]
As is, I get TUnit0013, since the method returns `IEnumerable<object[]>` and the test method has 6 parameters. Which makes sense given what the analyzer is trying to evaluate.

Since the method returns a mix of `IEnumerable<int>` and `IEnumerable<string>`, there's no type that I can use to match both, to force tunit to do the generic binding based on the inputs.

Attempting `public static IEnumerable<(IEnumerable<int>, Func<int, int>, Func<int, int>, Func<int, int, int>, IEqualityComparer<int>?, IEnumerable<KeyValuePair<int, int>>)> AggregateBy_TestData()` gives me a TUnit0006, since `IEnumerable<int>` doesn't match the unconstrained type `IEnumerable<TSource>`.

I'm going to put this one under the banner of "I'll find a different solution for my problem, but be aware of it for any future migrators from xunit/nunit".

**Reply by thomhurst** [reply]
The latter with tuples should work in theory. I'll take a look when I get a chance

**Reply by thomhurst** [reply]
Also, if you suppress the analyzer, does it work/run?

Curious if I need to fix two things or just the analyzer

**Reply by viceroypenguin** [reply]
I'm in the middle of a major refactor, so I can't retry on my code atm. When I'm closer to a place where it runs, I'll give it a try.

**Reply by thomhurst** [reply]
No worries. If I get some time tomorrow I'll try and reproduce it myself

**Reply by thomhurst** [reply, marked as answer]
@viceroypenguin  I had to make some changes to TUnit to support generic tests like this, but it works!
Example here: https://github.com/thomhurst/TUnit/blob/main/TUnit.TestProject/GenericMethodTests.cs 

```csharp
public class GenericMethodTests
{
    [Test]
    [MethodDataSource(nameof(AggregateBy_Numeric_TestData))]
    [MethodDataSource(nameof(AggregateBy_String_TestData))]
    public void AggregateBy_HasExpectedOutput<TSource, TKey, TAccumulate>(
        IEnumerable<TSource> source,
        Func<TSource, TKey> keySelector,
        Func<TKey, TAccumulate> seedSelector,
        Func<TAccumulate, TSource, TAccumulate> func,
        IEqualityComparer<TKey>? comparer,
        IEnumerable<KeyValuePair<TKey, TAccumulate>> expected) where TKey : notnull
    {
        var enumerable = source as TSource[] ?? source.ToArray();
        Console.WriteLine(string.Join(", ", enumerable));
        Console.WriteLine(string.Join(", ", enumerable.Select(keySelector)));
        Console.WriteLine(string.Join(", ", enumerable.Select(x => seedSelector(keySelector(x)))));
    }

    public static IEnumerable<(IEnumerable<int> source,
        Func<int, int> keySelector,
        Func<int, int> seedSelector,
        Func<int, int, int> func,
        IEqualityComparer<int>? comparer,
        IEnumerable<KeyValuePair<int, int>> expected)> AggregateBy_Numeric_TestData()
    {
        yield return (
            source: Enumerable.Range(0, 10),
            keySelector: x => x,
            seedSelector: x => 0,
            func: (x, y) => x + y,
            comparer: null,
            expected: Enumerable.Range(0, 10).ToDictionary(x => x, x => x)
        );
    }
    
    public static IEnumerable<(IEnumerable<string> source,
        Func<string, string> keySelector,
        Func<string, string> seedSelector,
        Func<string, string, string> func,
        IEqualityComparer<string>? comparer,
        IEnumerable<KeyValuePair<string, string>> expected)> AggregateBy_String_TestData()
    {
        yield return (
            source: ["Bob", "bob", "tim", "Bob", "Tim"],
            keySelector: x => x,
            seedSelector: x => "",
            func: (x, y) => x + y,
            comparer: null,
            expected: new Dictionary<string, string>(StringComparer.Ordinal)
            {
                { "Bob", "BobBob" },
                { "bob", "bob" },
                { "tim", "tim" },
                { "Tim", "Tim" },
            });
    }
}
```

**Reply by viceroypenguin** [reply]
Wow! You're incredible! I'll take advantage of that when I get back to SuperLinq tomorrow or Monday!  Thank you!
